File: backend/app/services/config_analyzer.py
from __future__ import annotations

import json
import re
from copy import deepcopy
from typing import Any

from jsonschema import ValidationError as JsonSchemaValidationError
from jsonschema import validate

from app.data.rules import OPTIMIZATIONS, RECOMMENDED_SETTINGS, SECURITY_RULES
from app.data.schema import OPENCODE_CONFIG_SCHEMA
from app.schemas.config import (
    ConfigAuditResult,
    ConfigDiffResult,
    DiffusionEntry,
    MissingSetting,
    Optimization,
    SchemaValidationError,
    SecurityIssue,
)


def _sanitize_json(text: str) -> str:
    """Strip BOM, zero-width spaces, and other invisible characters commonly
    introduced by copy-paste that break JSON parsers."""
    # Strip BOM
    text = text.lstrip("\ufeff")
    # Strip zero-width spaces, non-breaking spaces, left-to-right marks, etc.
    text = re.sub(r"[\u200b-\u200f\u2028-\u202f\ufeff\u00a0]", "", text)
    # Replace other C0 control chars (except \t, \n, \r which JSON allows)
    text = re.sub(r"[\x00-\x08\x0b\x0c\x0e-\x1f]", "", text)
    return text


def _strip_jsonc_comments(text: str) -> str:
    """Strip single-line and block comments from JSONC, plus trailing commas."""
    text = re.sub(r"/\*[\s\S]*?\*/", "", text)
    text = re.sub(r"//.*", "", text)
    text = re.sub(r",\s*([}\]])", r"\1", text)
    return text


def _format_parse_error(raw: str, e: json.JSONDecodeError) -> str:
    """Build a user-friendly parse error message showing the problematic area."""
    lines = raw.split("\n")
    if 1 <= e.lineno <= len(lines):
        line = lines[e.lineno - 1]
        start = max(0, e.colno - 20)
        end = min(len(line), e.colno + 20)
        snippet = line[start:end]
        marker_idx = e.colno - start - 1
        marker = " " * max(0, marker_idx) + "^"
        return (
            f"JSON parse error at line {e.lineno}, col {e.colno}: {e.msg}\n"
            f"  …{snippet}…\n"
            f"  {marker}"
        )
    return f"JSON parse error at line {e.lineno}, col {e.colno}: {e.msg}"


def parse_config(raw: str) -> tuple[dict[str, Any] | None, list[str]]:
    """Parse a raw JSON/JSONC string into a dict. Returns (parsed, errors)."""
    errors: list[str] = []
    cleaned = _sanitize_json(raw.strip())
    cleaned = _strip_jsonc_comments(cleaned)
    try:
        config = json.loads(cleaned)
        return config, errors
    except json.JSONDecodeError as e:
        errors.append(_format_parse_error(raw, e))
    # Retry with non-strict mode (allows control chars in strings)
    try:
        config = json.loads(cleaned, strict=False)
        return config, errors
    except json.JSONDecodeError as e:
        errors.append(_format_parse_error(raw, e))
    return None, errors


def _format_schema_errors(errors: list[JsonSchemaValidationError]) -> list[SchemaValidationError]:
    result: list[SchemaValidationError] = []
    for err in errors:
        path = err.json_path if err.json_path else "/"
        result.append(SchemaValidationError(path=path, message=err.message))
    return result


def analyze_config(config: dict[str, Any]) -> ConfigAuditResult:
    """Run schema validation, all security checks, missing-setting checks, and optimizations."""
    schema_errors: list[SchemaValidationError] = []
    security_issues: list[SecurityIssue] = []
    missing_settings: list[MissingSetting] = []
    optimizations: list[Optimization] = []

    try:
        validate(instance=config, schema=OPENCODE_CONFIG_SCHEMA)
    except JsonSchemaValidationError as e:
        schema_errors = _format_schema_errors(
            sorted(e.context, key=lambda x: x.json_path) if e.context else [e]
        )

    for rule in SECURITY_RULES:
        try:
            if rule["check"](config):
                security_issues.append(rule["issue"])
        except Exception:
            pass

    for rule in RECOMMENDED_SETTINGS:
        try:
            if rule["check"](config):
                missing_settings.append(rule["setting"])
        except Exception:
            pass

    for rule in OPTIMIZATIONS:
        try:
            if rule["check"](config):
                rec = rule["recommendation"]
                orig = config.get(rec["setting"])
                optimizations.append(
                    Optimization(
                        setting=rec["setting"],
                        original=orig,
                        recommended=rec["recommended"],
                        reason=rec["reason"],
                    )
                )
        except Exception:
            pass

    security_summary: dict[str, int] = {}
    for issue in security_issues:
        severity = issue.severity
        security_summary[severity] = security_summary.get(severity, 0) + 1

    opt_config = deepcopy(config)
    for rule in OPTIMIZATIONS:
        try:
            if rule["check"](config):
                rec = rule["recommendation"]
                opt_config[rec["setting"]] = rec["recommended"]
        except Exception:
            pass

    return ConfigAuditResult(
        is_valid_jsonc=True,
        validation_errors=[],
        schema_errors=schema_errors,
        security_issues=security_issues,
        security_summary=security_summary,
        missing_settings=missing_settings,
        optimizations=optimizations,
        optimized_config=opt_config,
    )


def compute_diff(original: dict[str, Any], modified: dict[str, Any]) -> ConfigDiffResult:
    """Compute a diff between two config dicts."""
    entries: list[DiffusionEntry] = []

    def _diff(
        a: Any,
        b: Any,
        path: str = "",
    ) -> None:
        if isinstance(a, dict) and isinstance(b, dict):
            all_keys = set(a.keys()) | set(b.keys())
            for key in sorted(all_keys):
                subpath = f"{path}.{key}" if path else key
                if key not in b:
                    entries.append(
                        DiffusionEntry(
                            path=subpath,
                            change="removed",
                            original=a[key],
                            modified=None,
                            description=f"Removed '{subpath}'",
                        )
                    )
                elif key not in a:
                    entries.append(
                        DiffusionEntry(
                            path=subpath,
                            change="added",
                            original=None,
                            modified=b[key],
                            description=f"Added '{subpath}'",
                        )
                    )
                else:
                    _diff(a[key], b[key], subpath)
        elif a != b:
            entries.append(
                DiffusionEntry(
                    path=path,
                    change="changed",
                    original=a,
                    modified=b,
                    description=f"Changed '{path}'",
                )
            )

    _diff(original, modified)
    return ConfigDiffResult(changes=entries)

File: frontend/src/lib/services/config-analyzer.ts
import Ajv2020 from "ajv/dist/2020"
import addFormats from "ajv-formats"
import OPENCODE_CONFIG_SCHEMA from "@/lib/data/schema"
import { SECURITY_RULES, RECOMMENDED_SETTINGS, OPTIMIZATIONS } from "@/lib/data/rules"
import type {
  ConfigAuditResult,
  ConfigDiffResult,
  DiffusionEntry,
  SchemaValidationError,
  SecurityIssue,
  MissingSetting,
  Optimization,
} from "@/types/config"

/* ------------------------------------------------------------------ */
/*  JSON / JSONC parsing                                              */
/* ------------------------------------------------------------------ */

function sanitize(text: string): string {
  return text
    .replace(/^\ufeff/, "")
    .replace(/[\u200b-\u200f\u2028-\u202f\ufeff\u00a0]/g, "")
    // eslint-disable-next-line no-control-regex
    .replace(/[\u0000-\u0008\u000b\u000c\u000e-\u001f]/g, "")
}

function stripComments(text: string): string {
  const result: string[] = []
  let i = 0
  const n = text.length
  let inString = false
  let stringChar = ""

  while (i < n) {
    const ch = text[i]!

    if (inString) {
      if (ch === "\\") {
        result.push(ch)
        i++
        if (i < n) result.push(text[i]!)
      } else if (ch === stringChar) {
        inString = false
        result.push(ch)
      } else {
        result.push(ch)
      }
      i++
      continue
    }

    if (ch === '"' || ch === "'") {
      inString = true
      stringChar = ch
      result.push(ch)
      i++
      continue
    }

    if (ch === "/" && i + 1 < n) {
      if (text[i + 1] === "/") {
        i += 2
        while (i < n && text[i] !== "\n") i++
        continue
      }
      if (text[i + 1] === "*") {
        i += 2
        while (i + 1 < n && !(text[i] === "*" && text[i + 1] === "/")) i++
        i += 2
        continue
      }
    }

    result.push(ch)
    i++
  }

  return result.join("").replace(/,\s*([}\]])/g, "$1")
}

export function parseConfig(raw: string): { config: Record<string, unknown> | null; errors: string[] } {
  const errors: string[] = []
  let cleaned = sanitize(raw.trim())
  cleaned = stripComments(cleaned)
  try {
    return { config: JSON.parse(cleaned) as Record<string, unknown>, errors }
  } catch (e) {
    errors.push(formatParseError(raw, e as SyntaxError))
  }
  try {
    return { config: JSON.parse(cleaned), errors }
  } catch (e) {
    errors.push(formatParseError(raw, e as SyntaxError))
  }
  return { config: null, errors }
}

function formatParseError(raw: string, e: SyntaxError): string {
  const match = /at line (\d+) column (\d+)/.exec(e.message)
  if (match) {
    const lineNo = parseInt(match[1]!, 10)
    const colNo = parseInt(match[2]!, 10)
    const lines = raw.split("\n")
    if (lineNo >= 1 && lineNo <= lines.length) {
      const line = lines[lineNo - 1]!
      const start = Math.max(0, colNo - 20)
      const end = Math.min(line.length, colNo + 20)
      const snippet = line.slice(start, end)
      const markerIdx = colNo - start - 1
      const marker = " ".repeat(Math.max(0, markerIdx)) + "^"
      return `JSON parse error at line ${lineNo}, col ${colNo}: ${e.message}\n  \u2026${snippet}\u2026\n  ${marker}`
    }
  }
  return `JSON parse error: ${e.message}`
}

/* ------------------------------------------------------------------ */
/*  Schema validation                                                  */
/* ------------------------------------------------------------------ */

let _ajv: Ajv2020 | null = null
let _validate: ReturnType<Ajv2020["compile"]> | null = null

function getValidator() {
  if (!_ajv) {
    _ajv = new Ajv2020({ strict: false })
    addFormats(_ajv)
    _validate = _ajv.compile(OPENCODE_CONFIG_SCHEMA)
  }
  return _validate!
}

function formatSchemaErrors(errors: { instancePath?: string; message?: string }[] | null | undefined): SchemaValidationError[] {
  if (!errors) return []
  return errors.map((e) => ({
    path: e.instancePath || "/",
    message: e.message || "Unknown schema error",
  }))
}

/* ------------------------------------------------------------------ */
/*  Config analysis                                                    */
/* ------------------------------------------------------------------ */

export function analyzeConfig(config: Record<string, unknown>): ConfigAuditResult {
  const schemaErrors: SchemaValidationError[] = []
  const securityIssues: SecurityIssue[] = []
  const missingSettings: MissingSetting[] = []
  const optimizations: Optimization[] = []

  try {
    const validate = getValidator()
    const valid = validate(config)
    if (!valid) schemaErrors.push(...formatSchemaErrors(validate.errors))
  } catch {
    /* schema validation is advisory */
  }

  for (const rule of SECURITY_RULES) {
    try {
      if (rule.check(config)) securityIssues.push({ ...rule.issue })
    } catch { /* ignore */ }
  }

  for (const rule of RECOMMENDED_SETTINGS) {
    try {
      if (rule.check(config)) missingSettings.push({ ...rule.setting })
    } catch { /* ignore */ }
  }

  for (const rule of OPTIMIZATIONS) {
    try {
      if (rule.check(config)) {
        const rec = rule.recommendation
        optimizations.push({
          setting: rec.setting,
          original: config[rec.setting] ?? null,
          recommended: rec.recommended,
          reason: rec.reason,
        })
      }
    } catch { /* ignore */ }
  }

  const securitySummary: Record<string, number> = {}
  for (const issue of securityIssues) {
    securitySummary[issue.severity] = (securitySummary[issue.severity] || 0) + 1
  }

  const optimizedConfig = structuredClone(config) as Record<string, unknown>
  for (const rule of OPTIMIZATIONS) {
    try {
      if (rule.check(config)) {
        const rec = rule.recommendation
        optimizedConfig[rec.setting] = rec.recommended
      }
    } catch { /* ignore */ }
  }

  return {
    is_valid_jsonc: true,
    validation_errors: [],
    schema_errors: schemaErrors,
    security_issues: securityIssues,
    security_summary: securitySummary,
    missing_settings: missingSettings,
    optimizations,
    optimized_config: optimizedConfig,
  }
}

/* ------------------------------------------------------------------ */
/*  Config diff                                                        */
/* ------------------------------------------------------------------ */

export function computeDiff(
  original: Record<string, unknown>,
  modified: Record<string, unknown>,
): ConfigDiffResult {
  const entries: DiffusionEntry[] = []

  function diff(a: unknown, b: unknown, path = ""): void {
    if (typeof a === "object" && a !== null && !Array.isArray(a) && typeof b === "object" && b !== null && !Array.isArray(b)) {
      const aObj = a as Record<string, unknown>
      const bObj = b as Record<string, unknown>
      const allKeys = [...new Set([...Object.keys(aObj), ...Object.keys(bObj)])].sort()
      for (const key of allKeys) {
        const subpath = path ? `${path}.${key}` : key
        if (!(key in bObj)) {
          entries.push({ path: subpath, change: "removed", original: aObj[key], modified: null, description: `Removed '${subpath}'` })
        } else if (!(key in aObj)) {
          entries.push({ path: subpath, change: "added", original: null, modified: bObj[key], description: `Added '${subpath}'` })
        } else {
          diff(aObj[key], bObj[key], subpath)
        }
      }
    } else if (a !== b) {
      entries.push({ path, change: "changed", original: a, modified: b, description: `Changed '${path}'` })
    }
  }

  diff(original, modified)
  return { changes: entries }
}

/* ------------------------------------------------------------------ */
/*  Full audit pipeline                                                */
/* ------------------------------------------------------------------ */

export function auditConfig(raw: string): ConfigAuditResult {
  const { config, errors } = parseConfig(raw)
  if (!config) {
    return {
      is_valid_jsonc: false,
      validation_errors: errors,
      schema_errors: [],
      security_issues: [],
      security_summary: {},
      missing_settings: [],
      optimizations: [],
      optimized_config: null,
    }
  }
  return analyzeConfig(config)
}
